lib/package_provider/repository.rb

```diff
@@ -1,72 +1,129 @@
 require 'tmpdir'
-require "open3"
+require 'open3'
 
+# Namespace that handles git operations for PackageProvider
 module PackageProvider
+  # Class for cloning remote or local git repositories
+  # and checkouting specified folders
   class Repository
-
-    attr_reader :repo_ur, :repo_folder, :local_git_repo_folder
+    attr_reader :repo_url, :repo_root
 
     class InvalidRepoPath < ArgumentError
     end
+    # rubocop:disable TrivialAccessors
+    def self.temp_prefix=(tp)
+      @temp_prefix = tp
+    end
 
-    def initialize(repo_url, local_git_repo_folder = nil)
-      @repo_url = repo_url
-      @repo_folder = Dir.mktmpdir('pprepo_')
-
-      if local_git_repo_folder
-        @local_git_repo_folder = local_git_repo_folder
-        raise InvalidRepoPath, 'Folder #{@local_git_repo_folder}' unless Dir.new(@local_repo_rot_folder).exists?
+    def self.temp_prefix
+      @temp_prefix || 'pp_repo_'
+    end
+    # rubocop:enable TrivialAccessors
+    def initialize(git_repo_url, git_repo_local_cache_root = nil)
+      if git_repo_local_cache_root
+        unless Dir.exist?(git_repo_local_cache_root)
+          fail InvalidRepoPath, "#{git_repo_local_cache_root} does not exists"
+        end
       end
 
-      clone!
+      @repo_url = git_repo_url
+      @repo_root = Dir.mktmpdir(self.class.temp_prefix)
+
+      init_repo!(git_repo_local_cache_root)
     end
 
-    def clone(dest_folder)
-      dest_folder
+    def logger
+      PackageProvider.logger
     end
 
-    def fetch()
-      fetch!
+    def clone(dest_dir, treeish, paths, use_submodules = false)
+      fail InvalidRepoPath, "Folder #{dest_dir} exists" if Dir.exist?(dest_dir)
+
+      logger.debug "clonning repo #{repo_root}: " \
+                   " [dest_dir: #{dest_dir.inspect}, " \
+                   "treeish: #{treeish.inspect}, " \
+                   "use_submodules: #{use_submodules.inspect}]"
+
+      begin
+        Dir.mkdir(dest_dir)
+        fetch(treeish)
+
+        fill_sparse_checkout_file(paths)
+        command = [
+          File.join(PackageProvider.root, 'lib', 'scripts', 'clone.sh'),
+          repo_root,
+          dest_dir,
+          treeish
+        ]
+        command << '--use-submodules' if use_submodules
+        run_command({}, command, chdir: repo_root)
+
```

Comment:
here, in the future, we will need `touch .package_provider_ready` to store that the target directory contains whole content.